src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -2556,56 +2493,28 @@ void InterpreterSelectQuery::executeFetchColumns(QueryProcessingStage::Enum proc
 
         /// Create optimizer with prepared actions.
         /// Maybe we will need to calc input_order_info later, e.g. while reading from StorageMerge.
-        if ((optimize_read_in_order || optimize_aggregation_in_order)
-            && (!query_info.projection || query_info.projection->complete))
+        if (optimize_read_in_order || optimize_aggregation_in_order)
```

Comment:
This if now would probably be easier to read if it was:
```
    if (optimize_read_in_order)
         ...
    else if (optimize_aggregation_in_order)
         ...
```

As both options are separate there is no need of doing (if (a || b) { if (a) else ...)